### PR TITLE
fix self.device access in DataParallel

### LINKED ISSUE #6413: self.device does not return the correct device in DataParallel
## 🐛 Bug

The` self.device` property does not get updated in the replicas of DataParallel.

## Please reproduce using the BoringModel

```python

### To Reproduce

import os

import torch
from torch.utils.data import Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        print(f"on device {batch.device.index} the value of self.device is {self.device}")
        output = self.layer(batch)
        return output.sum()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


if __name__ == '__main__':
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        max_epochs=1,
        weights_summary=None,
        accelerator="dp",
        gpus=2,
    )
    trainer.fit(model, train_data)

```

Output of this script is:

```
on device 0 the value of self.device is cuda:0
on device 1 the value of self.device is cuda:0
```

### LINKED ISSUE #6563: [BUG] `auto_move_data` does not work with `DataParallel`
## 🐛 Bug

In case your `forward` function is wrapped with `auto_move_data` it will not work with `DataParallel` because it will try to send the data to `self.device` which in dataparallel is always the main device.

i.e. the following won't work with `accelerator="dp"` (and probably also with "ddp"):
```python

class Module(pl.LightningModule):
     ...
     
     @auto_move_data
     def forward(x):
         ...

     def training_step(self, batch, batch_idx):
         x = self.forward(batch[0])
         ...
```

The error comes from this line: https://github.com/PyTorchLightning/pytorch-lightning/blob/b190403e282cbcb71147c7b618654476b08578a5/pytorch_lightning/core/hooks.py#L646 

`self.device` should probably be replaced by `torch.distributed.get_rank()` when `torch.distributed.is_available() and torch.distributed.is_initialized()`

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

DataParallel does not maintain state during the forward on the replicas. Therefore, self.device in e.g. training_step points to the root device no matter on which gpu.
We can fix this problem by taking the device from the input tensors, these are guaranteed to be on the right device.

The only exception is when the model doesn't get a batch with tensors in it. This is rare, and the best we can do is warn the user that self.device will not work.

Fixes #6413 
Fixes #6563

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Really Neat !

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]
